### ISSUE TITLE
Language Landing Pages

### ISSUE THREAD

**Issue by kytrinyx** [exercism/exercism.io#2299]
Mixolidia Gautreaux and Sarah Guermond have "dibsed" this feature.

----------

**TL;DR** It's hard to know what languages are available, what's available in each language, how to get started, and how to contribute.

The focus of this feature should initially be to provide insight about what's available on exercism. Information about contributing can come later.

----------

Suggested approach in order to be able to ship early and often:

**Expose existing documentation on homepage**

Add a section to the home page that lists each active language (the [x-api /tracks](http://x.exercism.io/tracks) endpoint provides this) and links to the current "getting started" page in the help site (`help.exercism.io/getting-started-with-$SLUG.html`).

**Add nominal landing pages**

* Create a single view that can be used for any of the languages.
* Call the [/tracks](http://x.exercism.io/tracks) endpoint, or perhaps the [/tracks/:language](http://x.exercism.io/tracks/ruby) endpoint so we can list the available problems in each language.
* Link to the "getting started" pages.
* Link to these landing pages from the home page, instead of linking to the help site.

**Deliver "getting started" content from Problems API**

* Each language repo would contain certain standardized files for things like installing dependencies and how to run the tests, and resources to get started learning the language.
* The ([Problems API](https://github.com/exercism/x-api) would have endpoints that deliver this content, e.g. `/docs/ruby`, `/docs/python`, etc.
* Add this content to the landing pages.

**Iterate and Improve**

Once the basics are in place we can add more sections to the documentation, e.g. how to contribute, resources about the language, etc.

----------

_Original issue and discussion follows._

At the moment it's quite difficult to see what languages are available, and for each language, what the track is like (which problems, what type of activity, how many people are involved, etc).

I'd like to design a flow that improves this.

Some ideas:

- from the homepage, we could have little 'cards' that have a language's logo along with the name.
- the top nav could also have a "languages" link that leads to a page that talks about which languages are available, encourages contributing more languages, and also has the same language 'cards' from the home page.
- each card would lead to a full page for that language

Each page could contain things like:

- The list of the problems in the language track, with links to see the readme and test suite
- instructions for getting started (installing dependencies, running tests, etc). We can detect your platform and show you the relevant instructions by default, with links to show the other instructions.
- links to where to post issues
- encouragement to contribute new exercises, port existing exercises from other languages, and improve existing exercises
- some key stats about activity in the track -- how many people have submitted code (ever), or activity this week

Eventually there might be other resources as well.

Some questions:

- What would the absolute simplest version of this be? It's still likely to be better than what we have.
- Is there any chance that we can get designers interested in contributing both in terms of designing the feature itself, but also things like artwork for the language logos? Designers generally don't do much open source, from what I can tell.

**Comment by pietvandongen** [comment]
Great idea, I totally agree that this is something that could add value! 

I always like breaking a (graphic) design problem up into smaller parts, creating requirement and use cases and only then think about the visual implications (wireframes, graphic design). I think it’s best to be as pragmatic as possible, so I only provide details that are necessary (I know, it’s still a long wall of text).

After agreeing on use cases, I think it’s best to start designing wireframes for them, one by one. When that’s agreed upon, think about the graphic design. After that, start thinking about implementation / technical details.

A proposal:

# Requirements

* **REQ-LP-1**: The user must be able to see what languages are available.
* **REQ-LP-2**: The user must be able to see what problems are available for a language track.
* **REQ-LP-3**: The user must be able to see how many people have solved a certain problem for a language track.
* **REQ-LP-4**: The user must be able to see how many people have been active for a language track since its start.
* **REQ-LP-5**: The user must be able to see how many people have been active during a certain period for a language track.
* **REQ-LP-6**: The user must be able to see the latest activity for a language track.
* **REQ-LP-7**: The user must be able to see the type of activity.
* **REQ-LP-8**: The user must be able to get instructions for a problem in a language track.
* **REQ-LP-9**: The user must be able to see the test suite for a problem in a language track.
* **REQ-LP-10**: The user must be able to get an overview of issues for a language track.
* **REQ-LP-11**: The user must be able to add an issue for a language track.
* **REQ-LP-12**: The user must be able to get instructions for getting started with a language track, based on their platform.
* **REQ-LP-13**: The user must be able to get instructions on how to add new exercises to the language track.
* **REQ-LP-14**: The user must be able to see what problems are not yet available to the language track.
* **REQ-LP-15**: The user must be able to see what problems are not to be implemented (foregone or deprecated) to the language track.
* **REQ-LP-16**: The user must be able to get an overview of the availability of the problem in other language track.
* **REQ-LP-17**: The user must be able to link to the same problem in another language track.
* **REQ-LP-18**: The user must be able to see how many problems are available for a language track.

# Use cases (brief format)

### Show list of available languages

#### Associated requirements

* REQ-LP-1
* REQ-LP-3
* REQ-LP-18

#### Description

The user can request an overview of available languages, including simple statistics per language.

### Show details for language track

#### Associated requirements

* REQ-LP-2
* REQ-LP-3
* REQ-LP-4
* REQ-LP-5
* REQ-LP-10
* REQ-LP-7
* REQ-LP-13

#### Description

The user can request an overview of the details for a single language track, including instructions to get started with the track, an overview of problems and their (simple) statistics and current activity.

### Show contribution page for a language track

#### Associated requirements

* REQ-LP-13
* REQ-LP-14
* REQ-LP-15

#### Description

The user can request an overview with contribution-related information for a language track, including a list of yet-to-be-implemented problems and the currently open issues for the language track. 

### Show contribution page for a problem

#### Associated requirements

* REQ-LP-16
* REQ-LP-17
* REQ-LP-10
* REQ-LP-11

#### Description

The user can request an overview of the implementation status of a certain problem in all language tracks. The user is also able to link to the implementation details.

### Show activity page for a language track

#### Associated requirements

* REQ-LP-5
* REQ-LP-6
* REQ-LP-7

#### Description

The user can request an overview with the activity for a language track, including the type of activity and the period during which the activity occurred. The user can also adjust this period.

### Show details of problem for language track

#### Associated requirements

* REQ-LP-1
* REQ-LP-18
* REQ-LP-8
* REQ-LP-9

#### Description

The user can request an overview of the details for a single problem in a language track, including the associated test suite.

**Comment by kytrinyx** [comment]
This is helpful, thanks.

I don't think that all of the things listed are necessarily requirements. The list of things that I posted was more of a braindump to get the conversation started. There might be other things that are requirements, and some of these things might not be useful at all.

I think that for the purposes of this feature, let's make a distinction between things that are about providing insight about what's available on the site versus information that is useful in order to contribute. The contribution bit can start out as simple as a sentence that says that we are always looking to improve the language tracks, and welcome contributions (link to contributing guide, link to language track repository). Then we can expand on that use case later.

The reason that I'm suggesting this is because I have four goals for the landing page, and only two of them are truly crucial for getting started with exercism

1. see what languages are available
2. see how to get started in a language
3. provide interesting information to make it more likely that you explore further
4. improve the language tracks by getting more people to contribute

3 and 4 are important but really unrelated to the first two (or each other). 4 in particular is secondary to the first two: if people are not participating in the exercises, it's not as important to improve them.


**Show list of available languages**

I think the associated requirements should be [1], and [18].

You also list REQ-LP-3 which I'm not sure about. [3] is about stats for a particular problem, and we're not showing problems at this point, just [18], the number of available problems. I think just the number of problems would be a great first step for the list of languages.

This could function as the very first, independently useful sub-feature, since we wouldn't even have to link to a landing page at first. We have the (really awkward but still useful) "getting started" pages in the help docs that we could link to.

I think that there are two use cases that should be prioritised; the rest are either improvements to the first two, or a separate concern:

1. Details for language track
2. Details of a problem

For (1) I think that this use case should be only raw data about a track itself (problems, how to get started), not activity. Activity is a plus, but a different use case.

**Contribution page for a track**

I like the idea of making this a separate page, though I think it's fine to start with just a small section on the language track page itself.

**Contribution page for a problem**

Very interesting idea, let's put it aside for later when digging into the contributions aspect of this more deeply.

**Activity page for a language track**

I'm not sure this needs a whole page. Let's start with just a section in the language track page.

**Comment by jwood803** [comment]
Would it make sense to put the getting started of each language in the landing page? Or maybe just a link to the current pages would suffice.

**Comment by kytrinyx** [comment]
I'd rather have the current getting started stuff move into these new landing pages. It might be a section that is hidden by default and can be expanded, and we might break it into multiple steps, but let's put it where people are going to be anyway.

I'm thinking let's put language-specific "getting started" content into text files in the language-specific repo and deliver it with the language track API (x-api).

**Comment by kytrinyx** [comment]
I've updated the first post in this discussion with a suggested order of implementation.

**Comment by kytrinyx** [comment]
It looks like we have the following sections in the language-specific help pages:

- setup (installing the language, installing dependencies, configuration)
- running the tests
- linters / style suggestions
- resources

I suggest that we add a `docs/` directory in each language repository that contains a file for each section (and later we can add more files):

- dependencies.md
- tests.md
- hints.md
- resources.md

**Comment by sguermond** [comment]
@mixolidia and I have been brainstorming what we want to do with this now that we have a very basic version of a landing page. 

We suggest a 3-page track for each language
1. Language info
2. Get started
3. How to use track/troubleshooting

**1. Language Info**
- Background on the language
- what it’s used for, brief history
- “You should learn this if...”
- short, 1-2 paragraphs max
- Button link to **2. Get Started**
- Link to **3. How to use track/troubleshooting**

**2. Get started**
- Exercism “getting started” reminder box at top: “You need to have CLI installed and we recommend homebrew for OSX/whatever for windows/whatever for linux. You also need a text editor. If you don’t have them installed here is where you should go first [link to Exercism getting started page]”
- Language “getting started” box underneath: List of packages required to run/compile/test the language
- Detailed install instructions (text + images, + video if we have it)
- Button link to **3. How to use track/troubleshooting**

**3. How to use track/troubleshooting**
- This would be a good place to have the readme file for each language
- how to run test suite
- link to that language’s tag in github user issues
- link to outside resources (want to learn more about this language? syntax guides, stackoverflow, codecademy intro tracks, etc.)

**Comment by kytrinyx** [comment]
This is fantastic, and I'd like to add two more possible sections/pages:

**4. how to contribute to the track** perhaps along with something about who is currently maintaining it, and who has contributed
**5. current activity** how many people have submitted exercises in the past day/week/month, and how lively are the conversations

**Comment by jwood803** [comment]
For 4 would it be worthwhile to include the main track maintainers in case folks need to get in touch with them? That may generate a lot of noise for those folks, though...

**Comment by kytrinyx** [comment]
Yeah, I think it would make sense, but we don't have to add their email addresses or usernames. We could explain how to tag the correct team if they want to raise an issue specifically for a particular repo.

**Comment by kytrinyx** [comment]
I'm going to close this issue, since we have the language landing pages (and the "about" section is in a pull request). I'll open a new issue with some thoughts around expanding them.

Thank you all :heart: :heart: :heart:
